### PR TITLE
Add endpoint to validate BLS Key - Closes #1287

### LINKED ISSUE #1287: Add endpoint to validate BLS Key
### Description

Allow users to validate their BLS Key.

### Acceptance Criteria

- The following endpoints are available
- Swagger specs are up-to-date

#### Endpoints
**HTTP**:  `POST /validator/validateBLSKey`
**RPC**:    `post.validator.validateBLSKey`

#### Request query params
Parameter | Type | Validation | Default | Comment
-- | -- | -- | -- | --
proofOfPossession | String | | (empty) | Required
blsKey | String |  | (empty) | Required

#### Response
```jsonc
{
  "data": {
    "isValid": true,
  },
  "meta": {
  },
  "links": {}
}
```

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1287

### How was it solved?
- The following endpoints are available
- [x] HTTP: POST /validator/validateBLSKey
- [x] RPC: post.validator.validateBLSKey
- [x] Swagger specs are up-to-date

### How was it tested?
- [x] Local
- [x] POST http://localhost:9901/api/v3/validator/validateBLSKey
`blsKey`: `b301803f8b5ac4a1133581fc676dfedc60d891dd5fa99028805e5ea5b08d3491af75d0707adab3b70c6a6a580217bf81`
`proofOfPossession`: `88bb31b27eae23038e14f9d9d1b628a39f5881b5278c3c6f0249f81ba0deb1f68aa5f8847854d6554051aa810fdf1cdb02df4af7a5647b1aa4afb60ec6d446ee17af24a8a50876ffdaf9bf475038ec5f8ebeda1c1c6a3220293e23b13a9a5d26`
- [x] rpc: ./moleculer_client.js indexer.validateBLSKey '{"blsKey":"b301803f8b5ac4a1133581fc676dfedc60d891dd5fa99028805e5ea5b08d3491af75d0707adab3b70c6a6a580217bf81","proofOfPossession":"88bb31b27eae23038e14f9d9d1b628a39f5881b5278c3c6f0249f81ba0deb1f68aa5f8847854d6554051aa810fdf1cdb02df4af7a5647b1aa4afb60ec6d446ee17af24a8a50876ffdaf9bf475038ec5f8ebeda1c1c6a3220293e23b13a9a5d26"}'

### CODE REVIEW

File: services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		data: {},
+		meta: {},
+	};
+
+	const { valid: isValid } = await requestConnector('validators_validateBLSKey', {
+		proofOfPossession,
+		blsKey,
+	});
+
+	response.data.isValid = isValid;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js:31]:
```suggestion
	const { valid } = await requestConnector('validators_validateBLSKey', {
		blsKey,
		proofOfPossession,
	});

	response.data.isValid = valid;
```

File: services/blockchain-indexer/methods/dataService/controllers/validator.js [hunk 2]
```diff
@@ -42,6 +43,19 @@ const getValidator = async params => {
...
 };
 
+const validateBLSKey = async params => {
+	const result = {
+		data: {},
+		meta: {},
+	};
+	const response = await business.validateBLSKey(params);
+	if (response.data) result.data = response.data;
+	if (response.meta) result.meta = response.meta;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/methods/dataService/controllers/validator.js:53]:
```suggestion
	};

	const response = await dataService.validateBLSKey(params);
	if (response.data) result.data = response.data;
	if (response.meta) result.meta = response.meta;
```

File: services/gateway/apis/http-version3/methods/validateBLSKey.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		validateBLSKeySchema[this.swaggerApiPath].post.tags = this.tags;
+		validateBLSKeySchema[this.swaggerApiPath].post.summary = 'Validates a BLS Key and Proof Of Possession pair';
+		validateBLSKeySchema[this.swaggerApiPath].post.description = getSwaggerDescription({
+			rpcMethod: this.rpcMethod,
+			description: 'Validates a BLS Key and Proof Of Possession pair',
+		});
+		validateBLSKeySchema[this.swaggerApiPath].post.parameters = [{ $ref: '#/parameters/validateBLSKeyParams' }];
+		validateBLSKeySchema[this.swaggerApiPath].post.responses = {
+			200: {
+				description: 'Returns a boolean representing the validity of BLS key and Proof Of Possesion pair',
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/methods/validateBLSKey.js:43]: Please always end the `summary` and `description` with a `.`.
```suggestion
		validateBLSKeySchema[this.swaggerApiPath].post.summary = 'Validates a BLS key against its corresponding Proof of Possession.';
		validateBLSKeySchema[this.swaggerApiPath].post.description = getSwaggerDescription({
			rpcMethod: this.rpcMethod,
			description: 'Validates a BLS key against its corresponding Proof of Possession.',
		});
		validateBLSKeySchema[this.swaggerApiPath].post.parameters = [{ $ref: '#/parameters/validateBLSKeyParams' }];
		validateBLSKeySchema[this.swaggerApiPath].post.responses = {
			200: {
				description: 'Returns a boolean representing the validity of the supplied BLS key and Proof of Possession.',
```

File: services/gateway/apis/http-version3/swagger/definitions/validators.json [hunk 4]
```diff
@@ -56,5 +56,44 @@
...
+            "proofOfPossession": {
+                "type": "string",
+                "example": "88bb31b27eae23038e14f9d9d1b628a39f5881b5278c3c6f0249f81ba0deb1f68aa5f8847854d6554051aa810fdf1cdb02df4af7a5647b1aa4afb60ec6d446ee17af24a8a50876ffdaf9bf475038ec5f8ebeda1c1c6a3220293e23b13a9a5d26",
+                "description": "Proof of Possession message"
+            },
+            "blsKey": {
+                "type": "string",
+                "example": "b301803f8b5ac4a1133581fc676dfedc60d891dd5fa99028805e5ea5b08d3491af75d0707adab3b70c6a6a580217bf81",
+                "description": "BLS Key"
+            }
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/swagger/definitions/validators.json:96]:
```suggestion
                "description": "Validity of the supplied BLS key and Proof of Possession.",
                "properties": {
                    "isValid": {
                        "type": "boolean",
                        "example": "true",
                        "description": "Returns a boolean representing the validity of the supplied BLS key and Proof of Possession."
                    }
                }
            },
            "meta": {
                "type": "object",
                "properties": {}
            }
        }
    },
    "blsKeyValidationParamEnvelope": {
        "type": "object",
        "properties": {
            "blsKey": {
                "type": "string",
                "example": "b301803f8b5ac4a1133581fc676dfedc60d891dd5fa99028805e5ea5b08d3491af75d0707adab3b70c6a6a580217bf81",
                "description": "The BLS key."
            },
            "proofOfPossession": {
                "type": "string",
                "example": "88bb31b27eae23038e14f9d9d1b628a39f5881b5278c3c6f0249f81ba0deb1f68aa5f8847854d6554051aa810fdf1cdb02df4af7a5647b1aa4afb60ec6d446ee17af24a8a50876ffdaf9bf475038ec5f8ebeda1c1c6a3220293e23b13a9a5d26",
                "description": "The Proof of Possession corresponding to the specified BLS key."
            }
```

File: services/blockchain-indexer/methods/dataService/validator.js [hunk 5]
```diff
@@ -25,4 +26,12 @@ module.exports = [
 			address: { optional: false, type: 'string' },
 		},
 	},
+	{
+		name: 'validateBLSKey',
+		controller: validateBLSKey,
+		params: {
+			proofOfPossession: { optional: false, type: 'string' },
+			blsKey: { optional: false, type: 'string' },
```
Comment (sameersubudhi) [inline services/blockchain-indexer/methods/dataService/validator.js:34]:
```suggestion
			blsKey: { optional: false, type: 'string' },
			proofOfPossession: { optional: false, type: 'string' },
```

File: services/blockchain-indexer/shared/dataService/business/validator/index.js [hunk 6]
```diff
@@ -15,6 +15,9 @@
  */
 const { getValidator } = require('./validatorInfo');
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/validator/index.js:17]:
```suggestion
```

File: services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+    "validateBLSKeyParams" : {
+        "name": "body",
+        "in": "body",
+        "required": true,
+        "description": "The request body must contain two properties:\n- proofOfPossession (required): proof of Possession message\n- blsKey (required): BLS Key",
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json:6]:
```suggestion
        "description": "The request body must contain two properties:\n- blsKey (required): The BLS key.\n- proofOfPossession (required): The Proof of Possession corresponding to the BLS key.",
```

File: services/blockchain-indexer/shared/dataService/business/index.js [hunk 8]
```diff
@@ -86,7 +86,10 @@ const {
 } = require('./dpos');
 
 const { getAuthAccountInfo } = require('./auth');
-const { getValidator } = require('./validator');
+const {
+	getValidator,
+	validateBLSKey,
+} = require('./validator');
 const { getEvents, getEventsByHeight } = require('./events');
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/index.js:93]:
```suggestion
const { getEvents, getEventsByHeight } = require('./events');
const { getValidator, validateBLSKey } = require('./validator');
```

File: services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const { requestConnector } = require('../../../utils/request');
+
+const validateBLSKey = async params => {
+	const { proofOfPossession, blsKey } = params;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js:19]:
```suggestion
	const { blsKey, proofOfPossession } = params;
```

Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js]: applied.

File: services/blockchain-indexer/shared/dataService/business/index.js (hunk 8 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/index.js]: applied.

File: services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json (hunk 7 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json]: applied.

File: services/blockchain-indexer/shared/dataService/business/validator/index.js (hunk 6 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/validator/index.js]: applied.

File: services/blockchain-indexer/shared/dataService/business/index.js (hunk 8 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/index.js]: applied.

File: services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js (hunk 1 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js]: applied.

File: services/gateway/apis/http-version3/swagger/definitions/validators.json (hunk 4 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/swagger/definitions/validators.json]: applied.

File: services/gateway/apis/http-version3/methods/validateBLSKey.js (hunk 3 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/validateBLSKey.js]: applied.

File: services/blockchain-indexer/methods/dataService/validator.js (hunk 5 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/methods/dataService/validator.js]: applied.

File: services/blockchain-indexer/methods/dataService/controllers/validator.js (hunk 2 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/methods/dataService/controllers/validator.js]: applied.

File: services/blockchain-indexer/methods/dataService/controllers/validator.js [hunk 10]
```diff
@@ -14,6 +14,7 @@
  *
  */
 // const dataService = require('../../../shared/dataService');
+const business = require('../../../shared/dataService/business');
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-indexer/methods/dataService/controllers/validator.js]: applied.

File: services/gateway/apis/http-version3/methods/validateBLSKey.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+module.exports = {
+	version: '2.0',
+	swaggerApiPath: '/validator/validateBLSKey',
+	httpMethod: 'POST',
+	rpcMethod: 'post.validator.validateBLSKey',
+	tags: ['Validator'],
+	params: {
+		proofOfPossession: { optional: false, type: 'string', pattern: regex.PROOF_OF_POSSESSION },
+		blsKey: { optional: false, type: 'string', pattern: regex.BLS_KEY },
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/methods/validateBLSKey.js:29]:
```suggestion
		blsKey: { optional: false, type: 'string', pattern: regex.BLS_KEY },
		proofOfPossession: { optional: false, type: 'string', pattern: regex.PROOF_OF_POSSESSION },
```

File: services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json [hunk 12]
```diff
@@ -0,0 +1,10 @@
+{
+    "validateBLSKeyParams" : {
+        "name": "body",
+        "in": "body",
+        "required": true,
+        "description": "The request body must contain two properties:\n- blsKey (required): The BLS key.\n- proofOfPossession (required): The Proof of Possession corresponding to the BLS key.",        "schema": {
+            "$ref": "#/definitions/blsKeyValidationParamEnvelope"
+        }
```
Comment (sameersubudhi) [inline services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json:8]:
```suggestion
        "description": "The request body must contain two properties:\n- blsKey (required): The BLS key.\n- proofOfPossession (required): The Proof of Possession corresponding to the BLS key.",
        "schema": {
            "$ref": "#/definitions/blsKeyValidationParamEnvelope"
        }
```

File: services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+const validateBLSKey = async params => {
+	const { blsKey, proofOfPossession } = params;
+
+	const response = {
+		data: {},
+		meta: {},
+	};
+
+	const { valid } = await requestConnector('validators_validateBLSKey', {
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js:26]: Please create an interface `validateBLSKey` in the connector and invoke the same.
```suggestion
	const response = {
		data: {},
		meta: {},
	};

	const { blsKey, proofOfPossession } = params;
	const { valid } = await requestConnector('validateBLSKey', {
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js (hunk 13 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/validator/validateBLSKey.js]: applied.

File: services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json (hunk 12 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/swagger/parameters/validateBLSKey.json]: applied.

File: services/gateway/apis/http-version3/methods/validateBLSKey.js (hunk 11 above)
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/validateBLSKey.js]: applied.

File: services/blockchain-connector/methods/validators.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ */
+const { validateBLSKey } = require('../shared/sdk/endpoints');
+const regex = require('../shared/utils/regex');
+
+module.exports = [
+	{
+		name: 'validateBLSKey',
+		controller: async ({ blsKey, proofOfPossession }) => validateBLSKey({
+			blsKey,
+			proofOfPossession }),
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-connector/methods/validators.js]: applied.

File: services/blockchain-connector/shared/sdk/endpoints.js [hunk 15]
```diff
@@ -294,6 +294,19 @@ const dryRunTransaction = async (transaction) => {
...
 };
 
+const validateBLSKey = async ({ blsKey, proofOfPossession }) => {
+	try {
+		const apiClient = await getApiClient();
+		const response = await apiClient._channel.invoke('validators_validateBLSKey', { blsKey, proofOfPossession });
+		return response;
+	} catch (err) {
+		if (err.message.includes(timeoutMessage)) {
+			throw new TimeoutException(`Request timed out when calling 'validators_validateBLSKey' with \nblsKey: ${blsKey}\nproofOfPossession: ${proofOfPossession}`);
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/endpoints.js]: applied.

File: services/blockchain-connector/shared/utils/regex.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const BLS_KEY = /^\b[a-fA-F0-9]{96}\b$/;
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/regex.js:16]: discussed offline.

**Review by sameersubudhi — APPROVED** [review]

**Review by nagdahimanshu — APPROVED** [review]
LGTM 👍

**Pull request merged**